### PR TITLE
MutableDocument equality check compares nodes of other document

### PR DESCRIPTION
**Pull request by britannio**
It was previously comparing `_nodes` with `nodes` but `nodes` is a getter that returns `_nodes` so they're identical and hence `const DeepCollectionEquality().equals(_nodes, nodes)` is always true.

This is my second PR related to misbehaving equality checks so would it be worth adding a few unit tests?

### CODE REVIEW

**Comment by matthew-carroll** [comment]
Unit tests sound like a good idea. Would you like to add some to this PR for this change?

**Comment by britannio** [comment]
> Unit tests sound like a good idea. Would you like to add some to this PR for this change?

Sure

**Comment by britannio** [comment]
@matthew-carroll Just added the tests!

File: super_editor/test/src/core/document_editor_test.dart [hunk 1]
```diff
@@ -21,5 +21,23 @@ void main() {
       // newNode exists at index 1
       expect(document.nodes.indexOf(newNode), 1);
     });
+
+    test('it is equal to a similar document with the same content', () {
```
Comment (matthew-carroll) [inline super_editor/test/src/core/document_editor_test.dart:25]: These test descriptions don't really describe what it means to be equal. Based on what I'm seeing, it looks like the description should read:

it is equal to another document when each content node is equal to the corresponding node in the other document

File: super_editor/test/src/core/document_editor_test.dart [hunk 2]
```diff
@@ -21,5 +21,23 @@ void main() {
...
+      final document2 = MutableDocument(nodes: [node]);
+
+      expect(document1 == document2, isTrue);
+    });
+
+    test('it is NOT equal to a similar document with different content', () {
+      final node = HorizontalRuleNode(id: '0');
+
+      final document1 = MutableDocument(nodes: [node]);
+      final document2 = MutableDocument(nodes: [node, node]);
```
Comment (matthew-carroll) [inline super_editor/test/src/core/document_editor_test.dart:38]: Tests should generally consist of situations that make sense in reality. There's no obvious situation where the same instance of a node would appear in a document, let alone back-to-back. This test should use a different node for the 2nd node in `document2`.

File: super_editor/test/src/core/document_editor_test.dart [hunk 3]
```diff
@@ -21,5 +21,23 @@ void main() {
...
+    test('it is equal to a similar document with the same content', () {
+      final node = HorizontalRuleNode(id: '0');
+
+      final document1 = MutableDocument(nodes: [node]);
+      final document2 = MutableDocument(nodes: [node]);
+
+      expect(document1 == document2, isTrue);
+    });
+
+    test('it is NOT equal to a similar document with different content', () {
```
Comment (matthew-carroll) [inline super_editor/test/src/core/document_editor_test.dart:34]: Referring to documents as "similar" is ambiguous. Similar how?

This test appears to test the following:
it is NOT equal to a document with the same starting nodes but with additional nodes at the end

File: super_editor/test/src/core/document_editor_test.dart [hunk 4]
```diff
@@ -21,5 +21,23 @@ void main() {
...
+
+    test('it is equal to a similar document with the same content', () {
+      final node = HorizontalRuleNode(id: '0');
+
+      final document1 = MutableDocument(nodes: [node]);
+      final document2 = MutableDocument(nodes: [node]);
+
+      expect(document1 == document2, isTrue);
+    });
+
```
Comment (matthew-carroll) [inline super_editor/test/src/core/document_editor_test.dart:33]: It's good to verify boundary cases and I think we're missing one here:
it is equal to another document when both documents are empty

File: super_editor/test/src/core/document_editor_test.dart [hunk 5]
```diff
@@ -21,5 +21,23 @@ void main() {
...
+    test('it is NOT equal to a similar document with different content', () {
+      final node = HorizontalRuleNode(id: '0');
+
+      final document1 = MutableDocument(nodes: [node]);
+      final document2 = MutableDocument(nodes: [node, node]);
+
+      expect(document1 == document2, isFalse);
+    });
```
Comment (matthew-carroll) [inline super_editor/test/src/core/document_editor_test.dart:41]: You've shown that a document with extra nodes will not be equal, but that means that an equality check could simply compare the `length` of the nodes and still pass these tests. It would be good to ensure that two lists with the same number of nodes are not equal when the nodes are not equal.

it is NOT equal to a document when corresponding nodes are NOT equal

**Review by matthew-carroll — CHANGES_REQUESTED** [review]
Here are some recommendations to improve the unit tests to better communicate the purpose of the tests, and to cover more boundary conditions.

**Comment by britannio** [comment]
Thanks for the feedback! Hopefully, I've addressed it all.

**Review by matthew-carroll — APPROVED** [review]
LGTM. Thanks.
